java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/connector/source/core/DbzChangeEventConsumer.java

```diff
@@ -211,16 +216,61 @@ var record = event.value();
             var response = respBuilder.build();
             outputChannel.put(response);
         }
+    }
 
-        committer.markBatchFinished();
+    public BlockingQueue<GetEventStreamResponse> getOutputChannel() {
+        return this.outputChannel;
     }
 
-    @Override
-    public boolean supportsTombstoneEvents() {
-        return DebeziumEngine.ChangeConsumer.super.supportsTombstoneEvents();
+    /**
+     * Commit the offset to the Debezium engine. NOTES: The input offset is passed from the source
+     * executor to here
+     *
+     * @param offset persisted offset in the Source state table
+     */
+    @SuppressWarnings("unchecked")
+    public void commitOffset(DebeziumOffset offset) throws InterruptedException {
+        // Although the committer is read/write by multi-thread, the committer will be not changed
+        // frequently.
+        if (currentRecordCommitter == null) {
+            LOG.info(
+                    "commitOffset() called on Debezium change consumer which doesn't receive records yet.");
+            return;
+        }
+
+        // only the offset is used
+        SourceRecord recordWrapper =
+                new SourceRecord(
+                        offset.sourcePartition,
+                        adjustSourceOffset((Map<String, Object>) offset.sourceOffset),
+                        "DUMMY",
+                        Schema.BOOLEAN_SCHEMA,
+                        true);
+        ChangeEvent<SourceRecord, SourceRecord> changeEvent =
+                EmbeddedEngineChangeEventProxy.create(null, recordWrapper, recordWrapper);
+        currentRecordCommitter.markProcessed(changeEvent);
+        currentRecordCommitter.markBatchFinished();
```

Comment:
```
/**
     * Contract passed to {@link ChangeConsumer}s, allowing them to commit single records as they have been processed
     * and to signal that offsets may be flushed eventually.
     */
    interface RecordCommitter<R> {

        /**
         * Marks a single record as processed, must be called for each
         * record.
         *
         * @param record the record to commit
         */
        void markProcessed(R record) throws InterruptedException;

        /**
         * Marks a batch as finished, this may result in committing offsets/flushing
         * data.
         * <p>
         * Should be called when a batch of records is finished being processed.
         */
        void markBatchFinished() throws InterruptedException;
```

The contract of `RecordCommitter` states that `markProcessed` must be called for **each** record and `markBatchFinished` should be called when the batch is processed. However, IIUC, in the current implementation, we will only call `markProcessed` per checkpoint. Also, the checkpoint doesn't necessarily align with the dbz batch boundary. Example:
```
dbz batch:
  record 1
  record 2
  record 3

// Prior to this PR
markProcessed(1);
markProcessed(2);
markProcessed(3);
markBatchFinished();

// After this PR, assume checkpoint happens after processing record 2 but before record 3
markProcessed(2);
markBatchFinished();

markProcessed(3);
...
markBatchFinished();

```

Will this be an issue?

Reply:
Thanks for the explanation. Indeed the same logic [here](https://github.com/apache/flink-cdc/blob/48ca8623bb8fa405adb56dbe505dbad10902db89/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-debezium/src/main/java/org/apache/flink/cdc/debezium/internal/DebeziumChangeConsumer.java#L83)